algosdk/atomic_transaction_composer.py

```diff
@@ -0,0 +1,529 @@
+from abc import ABC, abstractmethod
+import base64
+import copy
+from enum import IntEnum
+
+from algosdk import abi, error
+from algosdk.future import transaction
+
+# The first four bytes of an ABI method call return must have this hash
+ABI_RETURN_HASH = b"\x15\x1f\x7c\x75"
+
+
+class AtomicTransactionComposerStatus(IntEnum):
+    # BUILDING indicates that the atomic group is still under construction
+    BUILDING = 0
+
+    # BUILT indicates that the atomic group has been finalized,
+    # but not yet signed.
+    BUILT = 1
+
+    # SIGNED indicates that the atomic group has been finalized and signed,
+    # but not yet submitted to the network.
+    SIGNED = 2
+
+    # SUBMITTED indicates that the atomic group has been finalized, signed,
+    # and submitted to the network.
+    SUBMITTED = 3
+
+    # COMMITTED indicates that the atomic group has been finalized, signed,
+    # submitted, and successfully committed to a block.
+    COMMITTED = 4
+
+
+class AtomicTransactionComposer:
+    """
+    Constructs an atomic transaction group which may contain a combination of
+    Transactions and ABI Method calls.
+
+    Args:
+        status (AtomicTransactionComposerStatus): IntEnum representing the current state of the composer
+        method_dict (dict): dictionary of an index in the transaction list to a Method object
+        txn_list (list[TransactionWithSigner]): list of transactions with signers
+        signed_txns (list[SignedTransaction]): list of signed transactions
+        tx_ids (list[str]): list of individual transaction IDs in this atomic group
+    """
+
+    MAX_GROUP_SIZE = 16
+
+    def __init__(self) -> None:
+        self.status = AtomicTransactionComposerStatus.BUILDING
+        self.method_dict = {}
+        self.txn_list = []
+        self.signed_txns = []
+        self.tx_ids = []
+
+    def get_status(self):
+        """
+        Returns the status of this composer's transaction group.
+        """
+        return self.status
+
+    def get_tx_count(self):
+        """
+        Returns the number of transactions currently in this atomic group.
+        """
+        return len(self.txn_list)
+
+    def clone(self):
+        """
+        Creates a new composer with the same underlying transactions.
+        The new composer's status will be BUILDING, so additional transactions
+        may be added to it.
+        """
+        cloned = AtomicTransactionComposer()
+        cloned.method_dict = copy.deepcopy(self.method_dict)
+        cloned.txn_list = copy.deepcopy(self.txn_list)
+        for t in cloned.txn_list:
+            t.txn.group = None
```

Comment:
To me the name `clone` suggests an exact copy of the cloned object with all of its current state, but here the transaction groups are being nulled out. So maybe a more apropos name for the method would be something like `clone_skeleton()` or maybe just a dunder version that's not meant to be used externally such as `_clone()` ?

Reply:
I can see your point. I think we chose `clone` for the name so users can "clone" the transactions so they can build on top of it. You can't actually re-send the exact copy of the transactions because the blockchain will reject them while a transaction with the same hash is still in the pool (every transaction hash should be unique in the blockchain anyway). So I think the `clone` here sufficiently implies that it is being "cloned" for building, because cloning everything wouldn't have much use cases, because you'd have to break them down again and build them back up.

Reply:
👍